File: examples/appium/python/test_example.py
"""
Appium Python 示例 - 完整的自动化测试用例
"""

import pytest
from appium import webdriver
from appium.options.android import UiAutomator2Options
from appium.webdriver.common.appiumby import AppiumBy
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time
import random


class TestAppAutomation:
    """应用自动化测试示例"""
    
    @pytest.fixture(scope='function')
    def driver(self):
        """初始化驱动"""
        options = UiAutomator2Options()
        options.platform_name = "Android"
        options.device_name = "emulator-5554"
        options.app_package = "com.example.app"
        options.app_activity = ".MainActivity"
        options.automation_name = "UiAutomator2"
        options.no_reset = True
        options.new_command_timeout = 300
        
        driver = webdriver.Remote(
            'http://localhost:4723',
            options=options
        )
        
        yield driver
        driver.quit()
    
    def test_login_flow(self, driver):
        """测试登录流程"""
        # 等待登录页面加载
        wait = WebDriverWait(driver, 10)
        
        # 输入用户名
        username_field = wait.until(
            EC.presence_of_element_located(
                (AppiumBy.ID, "com.example:id/username")
            )
        )
        username_field.send_keys("test_user")
        
        # 输入密码
        password_field = driver.find_element(
            AppiumBy.ID, "com.example:id/password"
        )
        password_field.send_keys("test_password")
        
        # 点击登录按钮
        login_button = driver.find_element(
            AppiumBy.ID, "com.example:id/login_button"
        )
        login_button.click()
        
        # 验证登录成功
        success_message = wait.until(
            EC.presence_of_element_located(
                (AppiumBy.ID, "com.example:id/welcome_message")
            )
        )
        assert "欢迎" in success_message.text
    
    def test_swipe_gesture(self, driver):
        """测试滑动操作"""
        # 获取屏幕尺寸
        size = driver.get_window_size()
        width = size['width']
        height = size['height']
        
        # 向上滑动（从下到上）
        driver.swipe(
            start_x=width // 2,
            start_y=height * 0.8,
            end_x=width // 2,
            end_y=height * 0.2,
            duration=500
        )
        
        time.sleep(1)
        
        # 向下滑动（从上到下）
        driver.swipe(
            start_x=width // 2,
            start_y=height * 0.2,
            end_x=width // 2,
            end_y=height * 0.8,
            duration=500
        )
    
    def test_scroll_to_element(self, driver):
        """滚动查找元素"""
        # Android UiAutomator方式
        element = driver.find_element(
            AppiumBy.ANDROID_UIAUTOMATOR,
            'new UiScrollable(new UiSelector().scrollable(true))'
            '.scrollIntoView(new UiSelector().text("目标文本"))'
        )
        element.click()
    
    def test_handle_popup(self, driver):
        """处理弹窗"""
        try:
            # 等待弹窗出现（最多等待3秒）
            popup = WebDriverWait(driver, 3).until(
                EC.presence_of_element_located(
                    (AppiumBy.ID, "com.example:id/popup")
                )
            )
            
            # 点击关闭按钮
            close_button = driver.find_element(
                AppiumBy.ID, "com.example:id/close_button"
            )
            close_button.click()
        except:
            # 没有弹窗，继续执行
            pass
    
    def test_take_screenshot(self, driver):
        """截图示例"""
        # 保存截图
        driver.save_screenshot("screenshot.png")
        
        # 或者获取Base64编码
        screenshot_base64 = driver.get_screenshot_as_base64()
    
    def test_get_app_performance(self, driver):
        """获取应用性能数据"""
        # 获取应用性能数据（需要Appium 2.0+）
        performance_data = driver.get_performance_data(
            package_name='com.example.app',
            data_type='cpuinfo',
            data_read_timeout=1000
        )
        print(f"CPU信息: {performance_data}")
        
        # 获取内存信息
        memory_data = driver.get_performance_data(
            package_name='com.example.app',
            data_type='memoryinfo',
            data_read_timeout=1000
        )
        print(f"内存信息: {memory_data}")
    
    def test_switch_context(self, driver):
        """切换WebView上下文"""
        # 获取所有上下文
        contexts = driver.contexts
        print(f"可用上下文: {contexts}")
        
        # 切换到WebView
        for context in contexts:
            if 'WEBVIEW' in context:
                driver.switch_to.context(context)
                break
        
        # 在WebView中操作
        driver.find_element(AppiumBy.CSS_SELECTOR, "#element").click()
        
        # 切回Native
        driver.switch_to.context('NATIVE_APP')


class TestAdvancedFeatures:
    """高级功能示例"""
    
    @pytest.fixture(scope='function')
    def driver(self):
        """初始化驱动"""
        options = UiAutomator2Options()
        options.platform_name = "Android"
        options.device_name = "emulator-5554"
        options.app_package = "com.example.app"
        options.app_activity = ".MainActivity"
        
        driver = webdriver.Remote('http://localhost:4723', options=options)
        yield driver
        driver.quit()
    
    def test_clipboard_operations(self, driver):
        """剪贴板操作"""
        # 设置剪贴板内容
        driver.set_clipboard_text("Hello, World!")
        
        # 获取剪贴板内容
        clipboard_text = driver.get_clipboard_text()
        assert clipboard_text == "Hello, World!"
    
    def test_network_connection(self, driver):
        """网络连接控制"""
        from appium.webdriver.connectiontype import ConnectionType
        
        # 获取当前网络状态
        connection = driver.network_connection
        print(f"当前网络: {connection}")
        
        # 设置为飞行模式
        driver.set_network_connection(ConnectionType.AIRPLANE_MODE)
        
        time.sleep(2)
        
        # 恢复WiFi连接
        driver.set_network_connection(ConnectionType.WIFI_ONLY)
    
    def test_app_management(self, driver):
        """应用管理"""
        # 安装应用
        # driver.install_app('/path/to/app.apk')
        
        # 启动应用
        driver.activate_app('com.example.app')
        
        # 将应用置于后台（3秒）
        driver.background_app(3)
        
        # 检查应用状态
        app_state = driver.query_app_state('com.example.app')
        print(f"应用状态: {app_state}")
        
        # 终止应用
        driver.terminate_app('com.example.app')
        
        # 卸载应用
        # driver.remove_app('com.example.app')
    
    def test_file_operations(self, driver):
        """文件操作"""
        import base64
        
        # 推送文件到设备
        with open('local_file.txt', 'rb') as f:
            file_data = base64.b64encode(f.read()).decode('utf-8')
        
        driver.push_file('/sdcard/remote_file.txt', file_data)
        
        # 从设备拉取文件
        file_data_from_device = driver.pull_file('/sdcard/remote_file.txt')
        decoded_data = base64.b64decode(file_data_from_device)
        
        with open('pulled_file.txt', 'wb') as f:
            f.write(decoded_data)
    
    def test_execute_adb_command(self, driver):
        """执行ADB命令"""
        # 获取设备信息
        result = driver.execute_script('mobile: shell', {
            'command': 'getprop',
            'args': ['ro.product.model']
        })
        print(f"设备型号: {result}")
        
        # 发送广播
        driver.execute_script('mobile: shell', {
            'command': 'am',
            'args': ['broadcast', '-a', 'android.intent.action.BOOT_COMPLETED']
        })


if __name__ == '__main__':
    pytest.main([__file__, '-v', '--tb=short'])

File: examples/appium/typescript/test-example.ts
/**
 * Appium TypeScript 示例 - 完整的自动化测试用例
 */

import { remote, RemoteOptions } from 'webdriverio';
import type { Browser } from 'webdriverio';

// 测试配置
const capabilities: RemoteOptions = {
  hostname: 'localhost',
  port: 4723,
  path: '/',
  capabilities: {
    platformName: 'Android',
    'appium:deviceName': 'emulator-5554',
    'appium:appPackage': 'com.example.app',
    'appium:appActivity': '.MainActivity',
    'appium:automationName': 'UiAutomator2',
    'appium:noReset': true,
    'appium:newCommandTimeout': 300
  }
};

/**
 * 测试类：应用自动化测试
 */
class TestAppAutomation {
  private driver!: Browser;

  /**
   * 初始化驱动
   */
  async setup(): Promise<void> {
    this.driver = await remote(capabilities);
  }

  /**
   * 清理驱动
   */
  async teardown(): Promise<void> {
    await this.driver.deleteSession();
  }

  /**
   * 测试登录流程
   */
  async testLoginFlow(): Promise<void> {
    console.log('开始测试登录流程...');

    // 等待并输入用户名
    const usernameField = await this.driver.$('~username_field');
    await usernameField.waitForDisplayed({ timeout: 10000 });
    await usernameField.setValue('test_user');

    // 输入密码
    const passwordField = await this.driver.$('~password_field');
    await passwordField.setValue('test_password');

    // 点击登录按钮
    const loginButton = await this.driver.$('~login_button');
    await loginButton.click();

    // 验证登录成功
    const welcomeMessage = await this.driver.$('~welcome_message');
    await welcomeMessage.waitForDisplayed({ timeout: 10000 });
    const text = await welcomeMessage.getText();
    
    if (text.includes('欢迎')) {
      console.log('✅ 登录测试通过');
    } else {
      throw new Error('❌ 登录测试失败');
    }
  }

  /**
   * 测试滑动手势
   */
  async testSwipeGesture(): Promise<void> {
    console.log('测试滑动手势...');

    // 获取屏幕尺寸
    const { width, height } = await this.driver.getWindowSize();

    // 向上滑动
    await this.driver.touchAction([
      { action: 'press', x: width / 2, y: height * 0.8 },
      { action: 'wait', ms: 500 },
      { action: 'moveTo', x: width / 2, y: height * 0.2 },
      { action: 'release' }
    ]);

    await this.driver.pause(1000);

    // 向下滑动
    await this.driver.touchAction([
      { action: 'press', x: width / 2, y: height * 0.2 },
      { action: 'wait', ms: 500 },
      { action: 'moveTo', x: width / 2, y: height * 0.8 },
      { action: 'release' }
    ]);

    console.log('✅ 滑动测试完成');
  }

  /**
   * 滚动查找元素
   */
  async testScrollToElement(): Promise<void> {
    console.log('测试滚动查找元素...');

    // 使用 UiAutomator 滚动查找
    const element = await this.driver.$(
      'android=new UiScrollable(new UiSelector().scrollable(true))' +
      '.scrollIntoView(new UiSelector().text("目标文本"))'
    );
    
    await element.click();
    console.log('✅ 滚动查找元素成功');
  }

  /**
   * 处理弹窗
   */
  async testHandlePopup(): Promise<void> {
    console.log('测试处理弹窗...');

    try {
      const popup = await this.driver.$('~popup');
      const isDisplayed = await popup.isDisplayed();
      
      if (isDisplayed) {
        const closeButton = await this.driver.$('~close_button');
        await closeButton.click();
        console.log('✅ 弹窗已关闭');
      }
    } catch (error) {
      console.log('ℹ️ 没有弹窗');
    }
  }

  /**
   * 截图示例
   */
  async testTakeScreenshot(): Promise<void> {
    console.log('测试截图功能...');

    // 保存截图
    await this.driver.saveScreenshot('./screenshot.png');
    console.log('✅ 截图已保存');
  }

  /**
   * 切换上下文（WebView）
   */
  async testSwitchContext(): Promise<void> {
    console.log('测试切换上下文...');

    // 获取所有上下文
    const contexts = await this.driver.getContexts();
    console.log('可用上下文:', contexts);

    // 查找 WebView 上下文
    const webviewContext = contexts.find(ctx => ctx.includes('WEBVIEW'));
    
    if (webviewContext) {
      // 切换到 WebView
      await this.driver.switchContext(webviewContext);
      console.log(`已切换到: ${webviewContext}`);

      // 在 WebView 中操作
      const element = await this.driver.$('#element');
      await element.click();

      // 切回 Native
      await this.driver.switchContext('NATIVE_APP');
      console.log('已切换回 NATIVE_APP');
    }
  }
}

/**
 * 高级功能测试类
 */
class TestAdvancedFeatures {
  private driver!: Browser;

  async setup(): Promise<void> {
    this.driver = await remote(capabilities);
  }

  async teardown(): Promise<void> {
    await this.driver.deleteSession();
  }

  /**
   * 剪贴板操作
   */
  async testClipboard(): Promise<void> {
    console.log('测试剪贴板操作...');

    // 设置剪贴板内容
    await this.driver.setClipboard('Hello, World!', 'plaintext');

    // 获取剪贴板内容
    const clipboardText = await this.driver.getClipboard();
    
    if (clipboardText === 'Hello, World!') {
      console.log('✅ 剪贴板操作成功');
    } else {
      throw new Error('❌ 剪贴板操作失败');
    }
  }

  /**
   * 应用管理
   */
  async testAppManagement(): Promise<void> {
    console.log('测试应用管理...');

    const appPackage = 'com.example.app';

    // 启动应用
    await this.driver.activateApp(appPackage);
    console.log(`✅ 应用已启动: ${appPackage}`);

    // 将应用置于后台 3 秒
    await this.driver.background(3);
    console.log('✅ 应用已后台运行 3 秒');

    // 检查应用状态
    const appState = await this.driver.queryAppState(appPackage);
    console.log(`应用状态: ${appState}`);

    // 终止应用
    await this.driver.terminateApp(appPackage);
    console.log('✅ 应用已终止');
  }

  /**
   * 执行 Shell 命令
   */
  async testShellCommand(): Promise<void> {
    console.log('测试执行 Shell 命令...');

    // 获取设备型号
    const result = await this.driver.execute('mobile: shell', {
      command: 'getprop',
      args: ['ro.product.model']
    });

    console.log(`设备型号: ${result}`);
  }

  /**
   * 手势操作
   */
  async testGestures(): Promise<void> {
    console.log('测试手势操作...');

    const { width, height } = await this.driver.getWindowSize();

    // 缩放手势
    await this.driver.execute('mobile: pinchClose', {
      x: width / 2,
      y: height / 2,
      percent: 50
    });

    await this.driver.pause(1000);

    // 放大手势
    await this.driver.execute('mobile: pinchOpen', {
      x: width / 2,
      y: height / 2,
      percent: 50
    });

    console.log('✅ 手势操作完成');
  }
}

/**
 * 工具函数
 */
class TestUtils {
  /**
   * 随机延迟
   */
  static async randomDelay(min: number = 1000, max: number = 3000): Promise<void> {
    const delay = Math.floor(Math.random() * (max - min + 1)) + min;
    await new Promise(resolve => setTimeout(resolve, delay));
  }

  /**
   * 重试执行
   */
  static async retry<T>(
    fn: () => Promise<T>,
    maxAttempts: number = 3,
    delay: number = 1000
  ): Promise<T> {
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        return await fn();
      } catch (error) {
        if (attempt === maxAttempts) {
          throw error;
        }
        console.log(`尝试 ${attempt}/${maxAttempts} 失败，${delay}ms 后重试...`);
        await new Promise(resolve => setTimeout(resolve, delay));
      }
    }
    throw new Error('Should not reach here');
  }

  /**
   * 安全查找元素
   */
  static async findElementSafe(
    driver: Browser,
    selector: string,
    timeout: number = 5000
  ): Promise<WebdriverIO.Element | null> {
    try {
      const element = await driver.$(selector);
      await element.waitForDisplayed({ timeout });
      return element;
    } catch (error) {
      return null;
    }
  }
}

/**
 * 完整示例：自动化登录
 */
async function autoLoginExample(): Promise<void> {
  const driver = await remote(capabilities);

  try {
    console.log('开始自动化登录示例...');

    // 点击登录按钮
    const loginBtn = await TestUtils.findElementSafe(driver, '~login_button');
    if (loginBtn) {
      await loginBtn.click();
      console.log('✅ 点击登录按钮成功');
    } else {
      throw new Error('未找到登录按钮');
    }

    await TestUtils.randomDelay(500, 1500);

    // 输入用户名
    const usernameField = await driver.$('android.widget.EditText');
    await usernameField.setValue('username');
    await TestUtils.randomDelay(500, 1500);

    // 输入密码
    const fields = await driver.$$('android.widget.EditText');
    if (fields.length > 1) {
      await fields[1].setValue('password');
      await TestUtils.randomDelay(500, 1500);
    }

    // 点击确定
    const confirmBtn = await driver.$('~confirm_button');
    await confirmBtn.click();

    // 等待登录结果
    await driver.pause(3000);

    const successMsg = await TestUtils.findElementSafe(driver, '~success_message');
    if (successMsg) {
      console.log('✅ 登录成功！');
    } else {
      console.log('❌ 登录失败！');
    }
  } finally {
    await driver.deleteSession();
  }
}

/**
 * Page Object 模式示例
 */
class LoginPage {
  private driver: Browser;

  constructor(driver: Browser) {
    this.driver = driver;
  }

  // 定位器
  private get usernameField() {
    return this.driver.$('~username_field');
  }

  private get passwordField() {
    return this.driver.$('~password_field');
  }

  private get loginButton() {
    return this.driver.$('~login_button');
  }

  private get errorMessage() {
    return this.driver.$('~error_message');
  }

  /**
   * 执行登录
   */
  async login(username: string, password: string): Promise<void> {
    await (await this.usernameField).setValue(username);
    await (await this.passwordField).setValue(password);
    await (await this.loginButton).click();
  }

  /**
   * 获取错误信息
   */
  async getErrorMessage(): Promise<string> {
    const element = await this.errorMessage;
    return await element.getText();
  }

  /**
   * 判断是否登录成功
   */
  async isLoginSuccessful(): Promise<boolean> {
    try {
      const button = await this.loginButton;
      return !(await button.isDisplayed());
    } catch {
      return true;
    }
  }
}

/**
 * 使用 Page Object 的测试
 */
async function testWithPageObject(): Promise<void> {
  const driver = await remote(capabilities);

  try {
    const loginPage = new LoginPage(driver);

    // 测试有效登录
    await loginPage.login('valid_user', 'valid_pass');
    const isSuccess = await loginPage.isLoginSuccessful();
    console.log(isSuccess ? '✅ 登录成功' : '❌ 登录失败');

    // 测试无效登录
    await loginPage.login('invalid_user', 'wrong_pass');
    const errorMsg = await loginPage.getErrorMessage();
    console.log(`错误信息: ${errorMsg}`);
  } finally {
    await driver.deleteSession();
  }
}

/**
 * 主函数 - 运行所有测试
 */
async function main(): Promise<void> {
  console.log('🚀 开始运行 Appium TypeScript 测试...\n');

  // 基础测试
  const basicTests = new TestAppAutomation();
  await basicTests.setup();
  
  try {
    await basicTests.testLoginFlow();
    await basicTests.testSwipeGesture();
    await basicTests.testHandlePopup();
    await basicTests.testTakeScreenshot();
  } finally {
    await basicTests.teardown();
  }

  console.log('\n');

  // 高级测试
  const advancedTests = new TestAdvancedFeatures();
  await advancedTests.setup();
  
  try {
    await advancedTests.testClipboard();
    await advancedTests.testAppManagement();
    await advancedTests.testGestures();
  } finally {
    await advancedTests.teardown();
  }

  console.log('\n✅ 所有测试完成！');
}

// 导出供外部使用
export {
  TestAppAutomation,
  TestAdvancedFeatures,
  TestUtils,
  LoginPage,
  autoLoginExample,
  testWithPageObject
};

// 如果直接运行此文件
if (require.main === module) {
  main().catch(console.error);
}
